mgt-agenda should have protected render methods

shweaver-MSFT opened the pull request:
<!-- Review contributing guidelines before creating PRs -->
<!-- https://github.com/microsoftgraph/microsoft-graph-toolkit/blob/master/CONTRIBUTING.md -->

Contributes to #267 

### PR Type
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
- Feature
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Other... Please describe: -->

### Description of the changes

#267 describes how all components should expose protected render methods for easy extension of the components.

In this PR I've addressed only the `mgt-agenda` component. 

The render methods flow like so:
- `public render()`
  - `protected renderLoading()` - Render a loading state while the component loads.
  - `protected renderNoData(errorContext: object)` - Render an empty data state. 
  - `protected renderGroups(events: Event[])` - Sorts event data into groups and calls `renderHeader` before each group is rendered using `renderEvents`.
    - `protected renderHeader(header: string)`
  - `protected renderEvents(events: Event[])` - Renders a list of Event objects
    - `protected renderEvent(event: Event)` - Render an individual Event
      - `protected renderTitle(event: Event)` - Render Event title
      - `protected renderLocation(event: Event)` - Render Event location
      - `protected renderAttendees(event: Event)` - Render Event attendees
      - `protected renderOther(event: Event)` - Render additional Event metadata

### PR checklist
- [X] Project builds (`npm run build`) and changes have been tested in supported browsers (including IE11)
- [X] All public classes and methods have been documented
- [X] Added appropriate [documentation](https://github.com/microsoftgraph/microsoft-graph-docs/tree/master/concepts/toolkit)
- [X] License header has been added to all new source files (`npm run setLicense`)
- [X] Contains **NO** breaking changes

### Other information
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->

https://github.com/microsoftgraph/microsoft-graph-docs/pull/7276

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoftgraph/microsoft-graph-toolkit/pull/304)

nmetulev commented:
haven't done a full review, but have a suggestion regarding `no-data` vs `error` templates. I understand those as completely different:
* `no-data` - the call was succesful, but there were no events
* `error` - there was an error making the network call

Given this, I think both templates should co exist. If both are true (if there was an error, hence there is no data), then I would prioritize `error` over `no-data`, with `no-data` as a fallback if no `error` template is defined.

nmetulev commented:
IMO, it's ok to keep `renderGrouops` and `renderEvents` as protected and we should ensure that if the developer decided to overwrite those methods that the component continues to function with the developer provided structure.

shweaver-MSFT commented:
@nmetulev, I like your thoughts on the `error` vs `no-data` templates. I'll add support for both.

For the renderGroups/Events methods, I'll have to see how I can get that working. Stay tuned.

shweaver-MSFT commented:
@nmetulev, I made the `renderGroups`/`Events` methods protected, and didn't really need to change anything else. 

The only potential loss in functionality is that the `eventClick` method won't fire if `renderEvents` is overridden. But I think that's fine, since the overriding function will have context of the events. The control otherwise continues to function as expected.

shweaver-MSFT commented:
Also, I didn't really need to add support for the `error` template after all. `mgt-agenda` doesn't have any existing error logic. The support for `no-data` has been restored.

shweaver-MSFT commented:
I've updated the docs as well.
https://github.com/microsoftgraph/microsoft-graph-docs/pull/7267

nmetulev commented on src/components/mgt-tasks/mgt-tasks.ts:
```diff
@@ -1155,7 +1155,7 @@ export class MgtTasks extends MgtTemplatedComponent {
 
     const noPeopleTemplate = html`
-      <template data-type="no-data">
+      <template data-type="error">
```
this should be restored to `no-data`

shweaver-MSFT replied:
Good catch

nmetulev commented:
Here is the sample I used to test this - should we add something like this in examples for testing purposes?

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <script type="module" src="../../dist/es6/index.js"></script>
    <script type="module" src="../../dist/es6/mock/mgt-mock-provider.js"></script>
  </head>
  <body>
    <mgt-mock-provider></mgt-mock-provider>
    <my-agenda></my-agenda>

    <script type="module">
      import { MgtAgenda } from '../../dist/es6/index.js';
      import { html, css } from '../../node_modules/lit-element/lit-element.js';

      class MyAgenda extends MgtAgenda {
        static get styles() {
          return [
            MgtAgenda.styles,
            css`
              .my-event {
                color: red;
              }
            `
          ];
        }

        renderEvent(event) {
          return html`
            <div class="my-event">${event.subject}</div>
          `;
        }
      }

      window.customElements.define('my-agenda', MyAgenda);
    </script>
  </body>
</html>
```

shweaver-MSFT commented:
@nmetulev, I fixed the docs and the template issue in `mgt-tasks` 👍

Also, I love the sample you made! I added it to `samples/examples/agenda-extensions.html`
